### ISSUE TITLE
Diverse Learner: Exploring Diverse Supervision for Semi-supervised Object Detection

### ISSUE THREAD

**Issue by zhuoxiao-chen** [zhuoxiao-chen/please_read_paper_every_day#11]
### Paper Link
https://www.ecva.net/papers/eccv_2022/papers_ECCV/papers/136900631.pdf

## Motivation

- In traditional one pair of teacher-student model, EMA updates the teacher model in each training iteration, **the model weights of the teacher become extremely similar to those of the student**.
- This will lead to the fact that the prediction of the teacher model and the student model become nearly identical, which hinders the teacher model from digging information from the unlabeled data to supervise the student.
- A diverse learner framework, consisting of two-paired teacher-student models to **maintain the distinctiveness** of the teacher against the student.

### Semi-supervised Object Detection (SSOD)

$$
\begin{gathered}
L=L_s+\lambda L_u, \\
L_s=\frac{1}{N_s}\left(\sum_{i=1}^{N_s}\left(L_{c l s}\left(s_i, y_i\right)+\left(L_{r e g}\left(s_i, y_i\right)\right)\right)\right. \\
L_u=\frac{1}{N_u}\left(\sum_{i=1}^{N_u}\left(L_{c l s}\left(u_i, p_i\right)+\left(L_{r e g}\left(u_i, p_i\right)\right)\right),\right.
\end{gathered}
$$

#### Similarity Metric

$$
\text { Similarity }=\frac{1}{N_c} \sum_{i=1}^{N_c} \frac{M_{\text {tea }}^c \bigcap M_{\text {stu }}^c}{M_{\text {tea }}^c \bigcup M_{s t u}^c}
$$

- $N_c$ is the number of classes.
- $M_{\text {tea }}^c$ is obtained by aggregating all detection results of class $c$ produced by the teacher.
- The aggregation process is performed by setting the foreground area to 1 and background area to 0, thus generating a binary mask for each class.

This similarity metric measures the prediction consistency between the teacher and the student.

#### Similarity Evaluated with Increasing Training Iteration

<img width="400" alt="image" src="https://user-images.githubusercontent.com/50613939/204469339-fa6577c7-b5c4-4bfe-b4a4-80d5b43eb3b1.png">

#### Overview Framework
<img width="700" alt="image" src="https://user-images.githubusercontent.com/50613939/204472124-7b050d5d-ff09-45b8-bb62-872718355c10.png">

#### Unsupervised Loss

$$
\begin{gathered}
L_u=L_u^{s t u 1}+L_u^{s t u 2}, \\
L_u^{s t u 1}=L_u\left(f_s\left(S_1(U), \theta_{s t u 1}\right), P_{\text {tea } 2}\right) \\
L_u^{s t u 2}=L_u\left(f_s\left(S_2(U), \theta_{\text {stu2 }}\right), P_{\text {tea } 1}\right)
\end{gathered}
$$

#### Examples of the Pseudo Labels Generated
<img width="700" alt="image" src="https://user-images.githubusercontent.com/50613939/204718603-9b575874-cf17-4708-a836-e19324c2da4f.png">

Although some wrong pseudo labels in one teacher model, another teacher still keeps the opportunity for generating correct labels.

### Multi-Threshold Classification Loss

<img width="250" alt="image" src="https://user-images.githubusercontent.com/50613939/204720932-db1116de-092f-4e37-9b83-888a418f9d0b.png">

- For certain pseudo labels, we use one-hot label and cross entropy loss.
- For uncertain pseudo labels, we use soft label and **KL divergence loss**.

### Pseudo Label-Aware Erasing
<img width="700" alt="image" src="https://user-images.githubusercontent.com/50613939/204773631-01764ffc-a11b-4554-b70b-20738a88d4f4.png">

## Implementation 

- Faster-RCNN with FPN
- pre-trained on ImageNet and the parameters of detection heads randomly
- EMA update parameter $\alpha$ to 0.999
- For the multi-threshold classification loss, lower bound confidence threshold = 0.8, and upper bound confidence threshold = 0.9
- apply pseudo label-aware erasing strategy after 80000 iteration
- weak augmentation: random resize and horizontal flip
- strong augmentation: random erasing, rotation, color jittering, etc
- two teacher models both achieve high performance, we report the performance of the better teacher model

## Experiments 
<img width="800" alt="image" src="https://user-images.githubusercontent.com/50613939/204774959-44dc6257-a8ee-4713-a49c-fe4c45d0e1ff.png">

<img width="800" alt="image" src="https://user-images.githubusercontent.com/50613939/204775012-3a564fa2-cdc1-48fc-9eb5-639aabcc8094.png">

<img width="800" alt="image" src="https://user-images.githubusercontent.com/50613939/204775191-88a737f4-ba96-488c-8b25-59c1557b09fc.png">
